canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -133,42 +137,87 @@ private[apiserver] final class ApiPackageManagementService private (
+    optDar1.fold(Future.unit) { case (pkgId1, pkg1) =>
+        Future.fromTry(Typecheck.typecheckUpgrades((pkgId1, pkg1), optDar2))
+    }
+  }
+
+  private def validateUpgrade(upgradingDar: Dar[(Ref.PackageId, Ast.Package)])(implicit
       loggingContext: LoggingContextWithTrace
   ): Future[Unit] = {
-    val upgradingPackageId = upgradingPackage.main._1
+    val (upgradingPackageId, upgradingPackage) = upgradingDar.main
+    val optUpgradingDar = Some((upgradingPackageId, upgradingPackage))
     logger.info(s"Uploading DAR file for $upgradingPackageId in submission ID ${loggingContext.serializeFiltered("submissionId")}.")
-    upgradingPackage.main._2.metadata.flatMap(_.upgradedPackageId) match {
+    upgradingPackage.metadata.flatMap(_.upgradedPackageId) match {
       case Some(upgradedPackageId) =>
         for {
-          upgradedArchiveMb <- packagesIndex.getLfArchive(upgradedPackageId)
-          upgradedPackageMb <- Future.fromTry {
-            upgradedArchiveMb.traverse(Decode.decodeArchive(_))
-              .handleError(Validation.handleLfArchiveError)
-          }
-          _ <- {
-            logger.info(s"Package $upgradingPackageId upgrades package id $upgradedPackageId")
-            val upgradeCheckResult = Typecheck.typecheckUpgrades(upgradingPackage.main, upgradedPackageMb)
-            upgradeCheckResult match {
-              case Success(()) => {
-                logger.info(s"Typechecking upgrades for $upgradingPackageId succeeded.")
-                Future(())
-              }
-              case Failure(err: UpgradeError) => {
+          optUpgradedDar <- lookupDar(upgradedPackageId)
+          _ = logger.info(s"Package $upgradingPackageId upgrades package id $upgradedPackageId")
+          _ <- optUpgradingDar.fold {
+            logger.info(s"Typechecking upgrades for $upgradingPackageId failed with following message: CouldNotResolveUpgradedPackageId")
+            Future.failed(Validation.handleUpgradeError(upgradingPackageId, upgradedPackageId, UpgradeError("CouldNotResolveUpgradedPackageId")).asGrpcError)
+          } { _ =>
+            typecheckUpgrades(optUpgradingDar, optUpgradedDar).recoverWith {
+              case err: UpgradeError =>
                 logger.info(s"Typechecking upgrades for $upgradingPackageId failed with following message: ${err.message}")
                 Future.failed(Validation.handleUpgradeError(upgradingPackageId, upgradedPackageId, err).asGrpcError)
-              }
-              case Failure(err: Throwable) => {
+              case NonFatal(err) =>
                 logger.info(s"Typechecking upgrades failed with unknown error.")
                 Future.failed(err)
-              }
             }
           }
+          optMaximalDar <- maximalVersionedDar(upgradingDar.main)
+          _ <- typecheckUpgrades(optMaximalDar, optUpgradingDar)
+          optMinimalDar <- minimalVersionedDar(upgradingDar.main)
+          _ <- typecheckUpgrades(optUpgradingDar, optMinimalDar)
```

Comment:
This will fail if the optMinimalDar is not present, because typecheckUpgrades ignores if the first DAR is nonexistent and throws if the second DAR is nonexistent.

Reply:
The version of `typecheckUpgrades` I have defined is such that: whenever the minimal or maximal dar is None, then the type check returns with the future completed successfully. Its only when we have a minimal or maximal dar that the real type checking code is called.